[Project1/김영현] Notion 클로닝 프로젝트

loevray opened the pull request:
# 📌 5주차 프로젝트[Project1]

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->

## 요구 사항과 구현내용

https://notion-clone-vercel-gamma.vercel.app/

[스크린캐스트 2023년 10월 25일 21시 28분 41초.webm](https://github.com/prgrms-fe-devcourse/FEDC5-5_Project_Notion_VanillaJS/assets/87127340/54d0e84e-a852-4695-980d-980442dcf48b)

- [x] 바닐라 JS만을 이용해 노션을 클로닝합니다.
- [x] 기본적인 레이아웃은 노션과 같으며, 스타일링, 컬러값 등은 원하는대로 커스텀합니다.
- [x] 글 단위를 Document라고 합니다. Document는 Document 여러개를 포함할 수 있습니다.
- [x] 화면 좌측에 Root Documents를 불러오는 API를 통해 루트 Documents를 렌더링합니다.
- [x] Root Document를 클릭하면 오른쪽 편집기 영역에 해당 Document의 Content를 렌더링합니다.
- [x] 해당 Root Document에 하위 Document가 있는 경우, 해당 Document 아래에 트리 형태로 렌더링 합니다.
- [x] Document Tree에서 각 Document 우측에는 + 버튼이 있습니다. 해당 버튼을 클릭하면, 클릭한 Document의 하위 Document로 새 Document를 생성하고 편집화면으로 넘깁니다.
- [x] 편집기에는 기본적으로 저장 버튼이 없습니다. Document Save API를 이용해 지속적으로 서버에 저장되도록 합니다.
- [x] History API를 이용해 SPA 형태로 만듭니다.
- [x] 루트 URL 접속 시엔 별다른 편집기 선택이 안 된 상태입니다.
- [x] /documents/{documentId} 로 접속시, 해당 Document 의 content를 불러와 편집기에 로딩합니다.

## 추가 요구사항

- [x] 기본적으로 편집기는 textarea 기반으로 단순한 텍스트 편집기로 시작하되, 여력이 되면 div와 contentEditable을 조합해서 좀 더 Rich한 에디터를 만들어봅니다.
- [ ] 편집기 최하단에는 현재 편집 중인 Document의 하위 Document 링크를 렌더링하도록 추가합니다.
- [ ] 편집기 내에서 다른 Document name을 적은 경우, 자동으로 해당 Document의 편집 페이지로 이동하는 링크를 거는 기능을 추가합니다.
- [x] 그외 개선하거나 구현했으면 좋겠다는 부분이 있으면 적극적으로 구현해봅니다!(버튼 애니메이션, 새로고침시 저장..)

 <!-- ## ✅ 피드백 반영사항 지난 코드리뷰에서 고친 사항을 적어주세요. -->

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

- [x] 옵저버 패턴을 사용해보긴했는데, 제대로 사용한 것이 맞나요?
- [x] contenteditable 에디터 마크다운 처리방법이 궁금합니다(커서위치, 자음모음 분리관련)
- [x] `POST, PUT`등으로 서버의 데이터를 업데이트하면, 전역스토어가 이를 감지하고 데이터를 다시 fetching하여 사용하면 각 컴포넌트마다 props drilling과 상태관리가 편해질 것 같은데, 맞는 방법일까요?
- [x] `DocumentList`와 `DocumentItem`을 이용하여 재귀적 렌더링을 하였는데, 다른사람이 보았을때 단박에 이해하기 어려운 구조인 것 같습니다. 어떤 구조가 더 적절할까요?
- [x] 문서를 클릭하면, 문서 메뉴가 하이라이트 되게 만들었습니다. `documentlist`에서 렌더가 완료되었을때 호출하고(초기렌더링, 새로고침시), `window`에 이벤트리스너`popstate, route-change(커스텀이벤트)`를 등록하여 뒤로가기,앞으로가기,주소입력 을 처리하였습니다
괜찮은 방법인지 궁금합니다
- [x] 변수와 함수명이 제일 어렵습니다...ㅠㅠ`changeBackgroundSelectedDocument`처럼 함수명이 길어지면 어떻게 처리하시나요?
- [x] **svg**파일 처리가 궁금합니다. `export, import`를 처리하기위하여 각 js파일을 만들어 변수에 할당하였는데, **icon.svg**이런식으로 그냥 관리하는게 좋을까요?
- [x] 버튼상태를 로컬스토리지에 저장하는데, `{ key: dataset.id, value:  { isFoleded : boolean }}` vs `{ key: isFolded, value:{ 1:false,2:false, 3:true} }` 이런식으로 밸류에 datasetid : boolean 형식으로 저장하는게 좋을까요? 어떤방법이 좋은지 궁금합니다!
- [x] 재사용 가능한 컴포넌트는 최대한 의존성을 없애고 쪼개고, 조금 달라지는 컴포넌트면 상속해서 새로 제작하는게 좋을까요?
- [x] 항상 코드리뷰를 꼼꼼히 해주셔서 감사합니다

loevray committed:
Init: 초기생성

loevray committed:
Feat: Add documentItems
재귀적으로 DocumentList내부에서 렌더링되는 아이템이다

loevray committed:
Refator: App.js state logic
- 상태를 App.js에서 관리하고, createDoc,deleteDoc 둘 다 만들어 내부에 getDocuments함수 삽입
- App.js에서 제작한 메소드를 자식에게 뿌림. 자식들이 호출해도 getDocuments가 호출되어서 App상태가 변경됨

loevray committed:
Feat: Add document toggle button
If you click ▼ button, toggle children document

loevray committed:
Feat: Add routing
- 이름만 존재하던 라우트 함수 구현
- 아주 간단한 스스타일도 추가하였습니다

loevray committed:
Feat: Add documentAutoSave
- 본문 작성하는 기능 추가하였습니다
- use 'PUT' http-method

loevray committed:
Feat: Add router
- utils폴더에 router제작 => App과 멀리 떨어진 곳에서도 route이벤트 발생가능

loevray committed:
Refactor: change getDocument location
- Transferred from App.js to DocumentPage.js
- also, rename *getDocument => *fetchDocument

loevray committed:
Chore: Hide api
- hide api address and custom header

loevray committed:
Refacotr: routes, render
라우트와 렌더링 기능 대대적 리팩토링 진행이전 커밋

loevray committed:
Refactor: route,render done
- App내부 라우트, 렌더함수 리팩토링 완료하였습니다
- 전역 스토어 구현 및 옵저버 패턴 사용하여 상태 관리 해보겠습니다

loevray committed:
Style: nav styling
nav styling like notion

loevray committed:
Refactor: Button, Title Component
- abstract reusable components (Button, title)

loevray committed:
Feat: Add Icon
add svg file icon

loevray committed:
Refactor: recursive document render method
- move recursive document render method from DocumentList to DocumentItem
- DocumentList just a wrapped DocumentItem
- DocumentItem just documentNdoe, recursive render on oneself

loevray committed:
Refator: recursive render method in DocumentItem
- now not recursive render in DocumentItem on oneself
- recursive render DocumentList in DocumentItem
- and make wrapper node in DocumentItem (button, title...something in documentItem el)

loevray committed:
Design: modify Document item padding
- remove padding in documentItem
- add padding in documentItemWrapper

loevray committed:
Fix: remove documentPage initialState
remove assignmented variable documentPage initialState from App.js

loevray committed:
Feat: Add resizing nav bar
- now can resize nav bar like notion.so nav bar

loevray committed:
Feat: Add maxWidth in nav resizer

loevray committed:
Feat: Add simple observer object
- observer object has just three method
- subscribe(arguemnt is callback), unsubscribe, notify.

loevray committed:
Design: documentListHeader styling, add notion icon svg

loevray committed:
Feat: Add rich editor
- build a rich editor with contenteditable ndoe

loevray committed:
Fix: fix separted consonant and vowel in korean language
- restore caret only when markdown coverted
- if not markdown converted, can keep type

loevray committed:
Fix: contenteditable enter key bug
- fixed  missing contenteditalbe target error
- e.target === current editing node. so, don't need queryselector

loevray committed:
Fix: documentItem button fix

loevray committed:
Remove: remove useless commnets, remove useless folder

loevray committed:
Fix: nav resize-cursor bug fix
- not remove resize-cursor when mouseover in nav bug fix

paul01010 commented on src/page/NavPage.js:
```diff
@@ -0,0 +1,105 @@
+import Button from "../common/Button.js";
```
NavPage는 DocumentPage와함께 항상 렌더링되지않나요?
페이지(page)로 구분되는 기준이 궁금합니다~!

loevray replied:
큼지막한 친구들을 Page로 구분하였었는데...
지금 생각해보니 라우팅 별로 달라지는 친구들만 Page로 구분하는게 좋다고 생각되네요!
이름 `Nav`로 수정하고 폴더도 옮겼습니다 ㅎㅎ

paul01010 commented on src/App.js:
```diff
@@ -0,0 +1,31 @@
+    $target,
+  });
+  const routes = [{ path: "documents", component: DocumentPage }];
+  this.render = async (url) => {
```
render 함수가 실행된다는것은 initRouter를 타고 들어가보지않으면 이해하기 어렵습니다.

컴포넌트에서 명시적으로 render함수를 호출해주는것도 좋아보입니다.

loevray replied:
넵! 지금 보니 굳이 `router`에서 `DOMContentLoaded`를 부르지 않아도 될것 같아서 지우고 App에서 명시적으로 호출하도록 수정하였습니다!

paul01010 commented on src/common/Button.js:
```diff
@@ -0,0 +1,28 @@
+  const $button = document.createElement("button");
+  attributes.forEach((attribute) => {
+    const { name, value } = attribute;
+    if (name === "class") {
```
name에는 "class"말고 또 어떤 문자열이 올수있는걸까요?
제가 이어서 작업한다면, 클래스가 여러개잇는 케이스에서는 name에 어떤 값을 줘야하는걸까요?

코드리뷰나 이와 비슷한 커뮤니케이션 비용이 발생하지않으려면 name으로 올수있는 값들이 미리 정의되어있어야 좋습니다!

loevray replied:
attribute의 속성 이름은...각 버튼마다 필요한 속성이 다르다고 생각하기에 밖에서 주입하는 것이 옳다고 생각했습니다
클래스가 여러개 존재할때는 배열로 주어야한다고 주석에 적어놓겠습니다
혹시 더 나은 방법이 있다면 피드백 부탁드립니다!

paul01010 commented on src/component/DocumentItem.js:
```diff
@@ -0,0 +1,126 @@
+  $documentItem.classList.add("document-item");
+  $documentItemInner.classList.add("document-item-inner");
+  const storage = new Storage(window.localStorage);
+  const getChildDocuments = () => {
+    const childDocuments = $documentItem.querySelector(".document-children");
+    if (childDocuments) {
+      return childDocuments;
+    }
+    return undefined;
+  };
```
요소를 찾을수없으면 undefined를 반환하지않나요?
```suggestion
  const getChildDocuments = () => $documentItem.querySelector(".document-children");
```

loevray replied:
찾아보니 `null`을 반환합니다!
하지만 `falsy`한값이므로 예외처리를 지워주고 중괄호도 지워주었습니다
없을때는 어떤값이 반환되는지 잘 살펴보도록 하겠습니다!

paul01010 reviewed (commented):
고생하셨습니다~!
하나의 함수(컴포넌트)를 어떻게 더 작게 유지하고 역할을 분리해볼수있을지 고민해보면 좋을것같습니다.

> 옵저버 패턴을 사용해보긴했는데, 제대로 사용한 것이 맞나요?

네 잘 사용하셨습니다.

> contenteditable 에디터 마크다운 처리방법이 궁금합니다(커서위치, 자음모음 분리관련)

에디터는 만들기 어려운 굉장히 복잡한 기능을 가진 제품입니다. 그래서 대부분 라이브러리를 가져다가 입맛에맞게 수정하여 사용하죠.
커서위치는 Selection객체를 활용하구요. 자/모음 분리는 Intl.Collator를 사용하기도 합니다.


> POST, PUT등으로 서버의 데이터를 업데이트하면, 전역스토어가 이를 감지하고 데이터를 다시 fetching하여 사용하면 각 컴포넌트마다 props drilling과 상태관리가 편해질 것 같은데, 맞는 방법일까요?
 
 서버데이터를 업데이트한뒤, 다시 fetch 하여 데이터를 받아와야합니다.
 전역스토어가 업데이트를 감지하여 다시 fetch 하지 않습니다.
 스토어는 데이터의 보관함 역할만 할뿐입니다.
 
 > DocumentList와 DocumentItem을 이용하여 재귀적 렌더링을 하였는데, 다른사람이 보았을때 단박에 이해하기 어려운 구조인 것 같습니다. 어떤 구조가 더 적절할까요?
 
 재귀적인 렌더링은 적절해보입니다.
 다만 컴포넌트내부에서 많은 역할을 하고 있는것이 코드 가독성에 문제가 되고있습니다.
 컴포넌트는 view만 담당하고, 데이터의 보관, 변경하는 레이어의 분리, 함수의 분리 등이 필요해보입니다.
 (일주일이상 작업하셨고, 프로젝트가 거대해보이실텐데요. 지금 리팩토링하기엔 설계를 처음부터 뜯어고쳐야할수있습니다. 지금은 고민을 해보시고, 다음 프로젝트부터 조금씩 적용해보시 괜찮습니다. 새로운 학습과 과제를 진행하며 이 작업을 하기에는 시간이 부족할것입니다. )

> 문서를 클릭하면, 문서 메뉴가 하이라이트 되게 만들었습니다. documentlist에서 렌더가 완료되었을때 호출하고(초기렌더링, 새로고침시), window에 이벤트리스너popstate, route-change(커스텀이벤트)를 등록하여 뒤로가기,앞으로가기,주소입력 을 처리하였습니다
괜찮은 방법인지 궁금합니다

이 모든 행위들은 url이 변한다는 공통점이 있습니다.
url의 변화를 감지하면 더 간단하게 문서 메뉴의 하이라이팅을 구현할수있어보입니다. 

> 변수와 함수명이 제일 어렵습니다...ㅠㅠchangeBackgroundSelectedDocument처럼 함수명이 길어지면 어떻게 처리하시나요?

함수명이 길어진다는것은 너무많은 행위를 하고있다는말일수도 있습니다. 
함수내부 코드의 역할을 쪼개보세요.
그게 아니라 단어길이의 문제라면, 변수명에 쓰인 여러 단어의 조합을 의미하는 하나의 단어를 찾아내보세요.
원래 이게 젤 어렵습니다,,

> svg파일 처리가 궁금합니다. export, import를 처리하기위하여 각 js파일을 만들어 변수에 할당하였는데, icon.svg이런식으로 그냥 관리하는게 좋을까요?

네 svg로 관리해도 상관없습니다.

> 버튼상태를 로컬스토리지에 저장하는데, { key: dataset.id, value:  { isFoleded : boolean }} vs { key: isFolded, value:{ 1:false,2:false, 3:true} } 이런식으로 밸류에 datasetid : boolean 형식으로 저장하는게 좋을까요? 어떤방법이 좋은지 궁금합니다!

프로젝트마다 다를것입니다.
첫번째 방법은 id가 한정된 프로젝트에서 편리할거구요. 다만 Key가 2000만개면 문제가 생길수잇겟죠?
두번째 방법은 여러 버튼이 존재할때 쉽게 모든 데이터를 가져올수있겠습니다.
유리한걸 선택하면 됩니다. 무조건 좋은건없습니다.

> 재사용 가능한 컴포넌트는 최대한 의존성을 없애고 쪼개고, 조금 달라지는 컴포넌트면 상속해서 새로 제작하는게 좋을까요?
 
 좋은 질문이네요.
삼각형, 파란색이 칠해진 삼각형, 한쪽 변이 파란색인 삼각형, 모든꼭지점에 작은동그라미가 달린 삼각형, 무지개빛깔 삼각형등 여러 요구사항, 아주 조금 달라지는 변경 등이 있을수있겠죠.
상속을통해 해결할수도있겠지만, 섣부른 추상화는 시간이 지날수록 너무많은 분기처리가 생길수도있는 노릇입니다.
상황에 맞게 개발자가 선택해야할 문제라고 생각합니다.

OhWonJu commented on src/component/Editor.js:
```diff
@@ -0,0 +1,131 @@
+    newLine.setAttribute("placeholder", "내용을 입력하쎄용");
+    return newLine;
+  };
+  this.richEditorState = { content: null };
```
state 를 컴포넌트의 상단에 모을 것인지..
state 와 해당 state 를 핸들링하는 기능들을 기준으로 모을 것인지
에 대해서 개인적으로 매번 고민하는 부분입니다만..!
저는 state 들을 상단에 모아서 처음에 읽어나갈 때 "이러이러한 state 들로 유지되는 컴포넌트 이군." 하고 파악을 먼저하는 쪽으로 기울고 있는 것 같습니다. 영현님은 이에 대해서 어떻게 생각하시는지 궁금합니다.

loevray replied:
저도 상단에 모으는게 좋다고 생각합니다
구현에 급급한 나머지 본질을 잃어버렸네요..ㅎㅎ
상단에 모아보도록 하겠습니다!

OhWonJu reviewed (commented):
매번 과제를 수행함에 있어서 매우 깊은 고민을 하신게 느껴집니다.
저는 "그냥 과제니까..적당히 해야지, 리액트때 열심히 해야지.." 와 같은 생각을 가지고 최선을 다하지 않은 것 같아서 부끄럽고, 반성이 많이 되는 것 같습니다.

pkb9239 reviewed (commented):
잘 보고갑니다. 되게 코드가 더 저번보다 더 읽기 편해진거같습니다.  과제하시느라 수고 많으셨습니다. ! 항상 많이 배웁니다 !

Whoknow77 commented on src/utils/globalStore/Observer.js:
```diff
@@ -0,0 +1,30 @@
+class Observer {
```
유투브에서 몇번 보기만 했던 옵저버 패턴을 구현하신점이 인상적입니다.

loevray replied:
거창한 코드는 아니지만 막상 사용해보니 생각했던 것보단 어렵지 않더군요
괜히 겁만 잔뜩 먹고있었는데..ㅎㅎ
훈오님도 다음에 마음에 드는 패턴하나 사용해보시는건 어떨까요?

Whoknow77 commented on src/utils/api.js:
```diff
@@ -0,0 +1,21 @@
+export const request = async (url, options = {}) => {
+  try {
+    const res = await fetch(
+      `${API_END_POINT}${url.startsWith("/") ? url : `/${url}`}`,
```
오.. 로토님 초반 강의에서 본 것 같은데 전 벌써 까먹었었네요.. 좋은 코드습관인것 같습니다!

loevray replied:
제가 실수가 잦아서...굳이 안해도 되는 처리같긴 합니다ㅎㅎㅋㅋㅋ

Whoknow77 commented on src/utils/router.js:
```diff
@@ -0,0 +1,38 @@
+      onRoute();
+    }
+  });
+  //뒤로, 앞으로 가기시 라우팅(렌더링)
+  window.addEventListener("popstate", () => {
+    onRoute();
+  });
+  //첫화면, 새로고침시 라우팅(렌더링)
+  window.addEventListener("DOMContentLoaded", () => {
+    onRoute();
+  });
```
뒤로,앞으로 가기 뿐만아니라 처음 렌더링 할때도 라우팅을 고려하셨네요. 저는 컴포넌트에서 이 코드를 사용하였는데 따로 라우터로 뺄 수 있다는 것을 배워갑니다.

loevray replied:
결국 윈도우에 이벤트를 거는것이니 라우터로 빼도 적용이 되더군요! :smile:

Whoknow77 commented on index.html:
```diff
@@ -0,0 +1,13 @@
+    <link rel="stylesheet" href="/style.css"/> 
+    <title>Notion Clone</title>
+</head>
+<body id="root">
```
따로 body에 root id를 두신 이유가 있을까요??

loevray replied:
계속 보여야하는 `Nav`와 라우팅따라 보이거나 안보이는 `DocumentPage`때문에
`root`밑에 `Nav`를 append하여 항상 보이게하고
`main`밑의 컴포넌트들은 `replaceChildren`하여 화면을 다시 그리게 했습니다
지금 생각해보니 `display-none`으로하는게 통일성을 지켜주는 듯 하네요...!

Whoknow77 commented on src/page/NavPage.js:
```diff
@@ -0,0 +1,105 @@
+
+  const maxNavWidth = 500;
+  const minNavWidth = 250;
+  document.addEventListener("mousedown", (e) => {
```
진짜 노션처럼 드래그 앤 드롭으로 Nav바를 조절하는 기능을 구현하실 줄은 생각지도 못했습니다 ㄷㄷ
코드를 보니 직접 하나하나 이해하면서 계산하신 것 같은데 대단함을 느끼고 갑니다.. 처음부터 이 기능을 구현해야 겠다고 생각하신건가요? 그리고 유사한 기능을 이전에 구현해보신 적이 있으신가요?

loevray replied:
처음부터는 아니었지만 노션의 기능을 한두가지는 가져와야겠다고 생각해서 만들어봤습니다!
이런 기능은 처음 만들어보았지만 예전에 `canvas`api를 이용하여 그림판을 만들었던 방식과 유사해서 할 수 있던것 같습니다.
마우스의 좌표와 이런저런 좌표를 계산하기만 하면 됐습니다 :smile:

Whoknow77 commented on src/page/NavPage.js:
```diff
@@ -0,0 +1,105 @@
+export default function NavPage({ $target }) {
+  const $nav = document.createElement("nav");
+  //항상 존재해야하는 컴포넌트라서, 내부에서 타겟에 붙여주었다.
+  $target.prepend($nav);
```
맨날 append만 보다가 prepend는 처음보네요.. 어떤 기능을 하는건가요?

loevray replied:
첫자식으로 추가합니다!

Whoknow77 commented on src/page/ErrorPage.js:
```diff
@@ -0,0 +1,9 @@
+export default function ErrorPage({ $target }) {
+  const $errorPage = document.createElement("div");
+  $target.appendChild($errorPage);
+  this.render = () => {
+    $errorPage.innerHTML = `
+    컴포넌트가 존재하지 않습니다!
+    `;
+  };
+}
```
에러 페이지는 404 에러페이지에만 해당하는건가요??

loevray replied:
정확히는 `App.js`에서 라우팅시 존재하지 않는 컴포넌트의 주소를 참조했을때 에러페이지를 보여줍니다!

Whoknow77 commented:
영현님은 볼때마다 시간관리를 잘 하시는 것 같습니다. 10일의 짧은시간에도 노력의 흔적이 보입니다. 특히 옵저버 패턴을 구현한점과 컴포넌트를 분리할때 new 연산자를 자유롭게 사용하시는걸 보아 객체지향적인 고민도 들어가 있음을 느꼈습니다. 많이 배워갈 수 있었습니다!

loevray commented on src/common/Title.js:
```diff
@@ -0,0 +1,20 @@
+  };
+  $title.addEventListener("click", (e) => {
+    e.preventDefault();
+    //onTitleClick(e);
```
확인하였습니다!

loevray commented on src/component/DocumentItem.js:
```diff
@@ -0,0 +1,126 @@
+  this.state = initialState;
+  const $documentItem = document.createElement("div");
+  const $documentItemInner = document.createElement("div");
+  $documentItemInner.style.paddingLeft = `${depth > 0 ? depth * 10 : 10}px`;
```
넵! 매직넘버는 아예 걷어내는 쪽으로 리팩토링 해보겠습니다!

loevray committed:
Refactor: remove useless code(comment), magic number convert to constant

loevray commented on src/App.js:
```diff
@@ -0,0 +1,31 @@
+  new NavPage({
+    $target,
+  });
+  const routes = [{ path: "documents", component: DocumentPage }];
```
넵 한번 구조를 바꿔보도록 하겠습니다!
확실히 find보다 훨씬 좋은 성능을 기대할수있겠네요 ㅎ

loevray commented on src/component/Editor.js:
```diff
@@ -0,0 +1,131 @@
+  };
+
+  const keyDownHandler = (e) => {
+    //innerHTML수정하면 등록된 핸들러 날아가니까, 이벤트 위임 사용
```
엔터를치면 이벤트를 중지하고 div를 생성하여 다음줄로 붙여넣습니다.
백스페이스는 문자열이없으면 그 줄을 삭제하고 윗줄로 포커스를 넘기고
위,아래키는 줄을 이동합니다!
모두 cotenteditable관련 코드입니다.
한 함수에 너무 많은 기능이 담겨있었네요...ㅠㅠ

loevray commented on src/page/NavPage.js:
```diff
@@ -0,0 +1,105 @@
+    const documentsTree = await request("/documents");
+    documentList.setState(documentsTree);
+  };
+  const observer = Observer;
```
옵저버패턴의 옵저버를 가져오는 코드입니다!

loevray commented on src/common/Button.js:
```diff
@@ -0,0 +1,28 @@
+  attributes = [],
+  content = "",
+  onClick = () => {},
+}) {
```
감사합니다! 태어나서 처음으로 재사용성 높은 컴포넌트를 만들어본 것 같네요 ㅎㅎ

loevray commented on src/component/DocumentList.js:
```diff
@@ -0,0 +1,52 @@
+  if (depth > 0) {
+    $documentList.classList.add("document-children", "display-none");
+  }
+  const changeBackgroundSelectedDocument = () => {
```
넵! 한번 이름을 잘줄여보겠습니다
 작명하는게 제일 어려운 것 같습니다ㅠㅠㅋㅋ

loevray committed:
Rename: DocumentList method rename
changeBackgroundSelectedDocument to highlightSelectedDocument

loevray commented on src/utils/globalStore/Observer.js:
```diff
@@ -0,0 +1,30 @@
+    this.subscribers = [];
+  }
+  subscribe(observerCallback) {
+    this.subscribers.push(observerCallback);
```
!!!!
전혀 생각지도 못했던건데 감사합니다...
`set`을 사용하여 같은 콜백을 없애보겠습니다!

loevray committed:
Refactor: Observer subscribers type change
Array to Set

loevray commented on src/page/ErrorPage.js:
```diff
@@ -0,0 +1,9 @@
+export default function ErrorPage({ $target }) {
```
넵! 라우팅 할시 주소가 없다면 에러페이지를 띄우게 해봤습니다만...페이지가 하나라 사실 상관없어 보입니다 ㅋㅋ

loevray commented on src/page/NavPage.js:
```diff
@@ -0,0 +1,105 @@
+      method: "POST",
+      body: JSON.stringify(body),
+    });
+    console.log(response);
```
넵 불필요한 주석과 콘솔로그는 제거하겠습니다!

loevray commented on src/utils/router.js:
```diff
@@ -0,0 +1,38 @@
+    const { nextUrl, callback } = e.detail;
+    if (nextUrl) {
+      history.pushState(null, null, nextUrl);
+      if (callback) {
+        callback();
+      }
```
와우 이건 충격이네요!
적용 후 옵셔널 체이닝에 대해 보다 더 자세히 알아보겠습니다

loevray committed:
Refactor: change router dataStructure
Array to Map

loevray commented on src/App.js:
```diff
@@ -0,0 +1,31 @@
+        $target: $app,
+        initialState: { id: pathData },
+      });
+    } else new ErrorPage({ $target: $app });
```
넵! 그렇게 바꿔보겠습니다!
한가지 여쭙고 싶은것이 있는데...

```js
  const routes = new Map();
  routes.set("documents", DocumentPage);

  this.render = async (url) => {
    const path = url ?? window.location.pathname;
    const [, pathname, pathData] = path.split("/");
    if (pathname === "") {
      //메인이면 메인 비워주기
      $app.innerHTML = "";
      return;
    }
    const component = routes.get(pathname) || ErrorPage;
    //라우팅 되는 직계 자식들은 replaceChildren으로 깜빡임 방지..
    new component({
      $target: $app,
      initialState: { id: pathData },
    });
  };
```

위의 `routes.set`의 컴포넌트를 넘겨주는 부분에 프롭스를 같이 넘겨주고싶은데, 하필 프롭스가 path에서 뽑아오는 데이터입니다.
따라서 `this.render`내부에서 `pathdata`를 뽑아 넘겨주는 형식인데, 라우팅하는 페이지가 늘어난다면 `new component`부분의 프롭스를 주는 부분을 분기처리해야 될 것 같습니다
분기처리를 하지 말고 `routes.set`을 할때 넘겨주는 방식을 사용할순 없을까요?

loevray committed:
Rename: NavPage to Nav

loevray commented on src/svg/arrowIcon.js:
```diff
@@ -0,0 +1,9 @@
+const arrowIconSvg = `
```
저도 어떻게할까 고민하다가 모듈화를 시키면 좋을것 같아 분리하였습니다
svg파일 자체로 보관 후 `img src="..."`이렇게 불러오는 방법도 있더라고요!

loevray committed:
Refactor: getChildDocuments method

loevray committed:
Refactor: router, App.render
- remove DOMContetLoaded event in router.js(useless)
- explicit call render method in App.js

loevray committed:
Style: remove assignment observer
Use imported Observer not assignment to variables

loevray committed:
Refactor: separate logic inside keyDownHandler
- separate enterKeyDown, backSpaceKeyDown function

loevray committed:
Style: pull up editor states

loevray committed:
Feat: Add util of getTag, isEqual, validateState and Add Component class
- getTag is getType Of value
- isEqual is check two values are the same
- validateState check initialState and nextState type are the same
- Component class is default class of Component

loevray committed:
Refacotr: change logic each other render, createElement in Component class
- createElement => createTemplate

loevray committed:
Refacotr: Add coditional wrapper in Component class
- if has tagName, make a wrapper else make no wrapper

loevray committed:
Feat: Add Route class, Router class

loevray committed:
Refacotr: diverted this.wrapper in Component class
- if has no wrapper, no action related to this.wrapper

loevray committed:
Fix: this binding in Router.js
Modify this binding arrow function when transferring callback function from router's addRouteEvent message to initRouter

loevray committed:
Commnet: Add comment in Router.js

loevray committed:
Fix: env.js

loevray committed:
Remove: globalStore, observer move to observer folder

loevray committed:
Feat: Add getDeepCopy method, Observer => Observable
- Obsever class separate Observable, Observer
- Observable done, now make Observer

loevray committed:
Feat: Add combineReducers, remove Observer class

loevray committed:
Feat: Add document module(reducer) with Ducks Pattern

loevray committed:
Feat: Add globalStore in main

loevray committed:
Fix: dispatch method in createStore
dispatch change state

loevray committed:
Refactor: use getDeepCopy function in dispatch method

loevray committed:
Test: change getDocuments method to redux reducer

loevray committed:
Fix: resize nav cursor bug fix

loevray committed:
Refactor: Add fragment node in handleRoute method
Initialize the fragment and attach it to the fragment without attaching it directly to the target node

loevray committed:
Feat: Add prepare method in Component class

loevray committed:
Refactor: depth parameter move in props at documentList

loevray committed:
Fix: stop event bubbling in documentItem class (click event)

loevray committed:
Feat: Add fetchCurrentDocument thunk function in documentsDuck.js

loevray committed:
Fix: Flickering fix
if routing component, use replaceChildren instead appendChild

loevray committed:
Refactor: constructor function => class (documentPage)

loevray committed:
Fix: useSelector, isEqaul, Observe
- useSelector can subscribe state
- isEqaul => isEqual. and fixed compare object with Object.keys.length
- remove Observable, add Observe
- Observe function each property has getter, setter

loevray committed:
Fix: re-rendered based on the number of child and parent documents
- Add unobserve function in Observe

loevray committed:
Refactor: Title constructor function => Title class

loevray committed:
Refactor: DocumentListHeader modified cosntructor function => class

loevray committed:
Refactor: button component modified constructor function => class

loevray committed:
Refactor: NoSubDocument modified contructor function => class

loevray committed:
Feat: add based 'id' nomarlization function

loevray committed:
Fix: re-render bug fix, add unmount, beforeunmount method in documentPage component

loevray committed:
Refactor: Editor component contructor function => class

loevray committed:
Fix: rotate arrow icon method fix
rotateSvg method has parameter it selector

loevray committed:
Fix: re-render bug fix
